scripts/get-py-version.go

```diff
@@ -13,11 +14,93 @@ func main() {
+	releaseVersionRegex = regexp.MustCompile(
+		`^v(?P<version>\d+\.\d+\.\d+)(?P<dirty>\+dirty)?$`)
+	rcVersionRegex = regexp.MustCompile(
+		`^v(?P<version>\d+\.\d+\.\d+)-rc\.(?P<rcN>\d+)(?P<dirty>\+dirty)?$`)
+	betaVersionRegex = regexp.MustCompile(
+		`^v(?P<version>\d+\.\d+\.\d+)-beta\.(?P<betaN>\d+)(?P<dirty>.dirty)?$`)
+	alphaVersionRegex = regexp.MustCompile(
+		`^v(?P<version>\d+\.\d+\.\d+)-alpha\.(?P<time>\d+)\+(?P<gitInfo>g[a-z0-9]+)(?P<dirty>.dirty)?$`)
+	devVersionRegex = regexp.MustCompile(
+		`^v(?P<version>\d+\.\d+\.\d+)-dev\.(?P<time>\d+)\+(?P<gitInfo>g[a-z0-9]+)(?P<dirty>.dirty)?$`)
+)
+
+// pyPiVersionFromNpmVersion returns a PEP-440 compliant version for a given semver version. This method does not
+// support all possible semver strings, but instead just supports versions that we generate for our node packages.
+//
+// NOTE: We do not include git information in the generated version (even within the local part, which PEP440 would
+// allow) because we publish dev packages to PyPI, which does not allow local parts. Instead, we only add a local part
+// when the build is dirty (which has the nice side effect of preventing us from publishing a build from dirty bits).
+func pyPiVersionFromNpmVersion(s string) (string, error) {
+	var b bytes.Buffer
+
+	switch {
+	case releaseVersionRegex.MatchString(s):
+		capMap := captureToMap(releaseVersionRegex, s)
+		mustFprintf(&b, "%s", capMap["version"])
+		if capMap["dirty"] != "" {
+			mustFprintf(&b, "+dirty")
+		}
+		return b.String(), nil
+	case rcVersionRegex.MatchString(s):
+		capMap := captureToMap(rcVersionRegex, s)
+		mustFprintf(&b, "%src%s", capMap["version"], capMap["rcN"])
+		if capMap["dirty"] != "" {
+			mustFprintf(&b, "+dirty")
+		}
+		return b.String(), nil
+	case betaVersionRegex.MatchString(s):
+		capMap := captureToMap(betaVersionRegex, s)
+		mustFprintf(&b, "%sb%s", capMap["version"], capMap["betaN"])
+		if capMap["dirty"] != "" {
+			mustFprintf(&b, "+dirty")
+		}
+		return b.String(), nil
+	case alphaVersionRegex.MatchString(s):
+		capMap := captureToMap(alphaVersionRegex, s)
+		mustFprintf(&b, "%sa%s", capMap["version"], capMap["time"])
+		if capMap["dirty"] != "" {
+			mustFprintf(&b, "+dirty")
+		}
+		return b.String(), nil
+	case devVersionRegex.MatchString(s):
+		capMap := captureToMap(devVersionRegex, s)
+		mustFprintf(&b, "%s.dev%s", capMap["version"], capMap["time"])
+		if capMap["dirty"] != "" {
+			mustFprintf(&b, "+dirty")
+		}
+		return b.String(), nil
+	}
+
+	return "", fmt.Errorf("can not parse version string '%s'", s)
+}
+
+func captureToMap(r *regexp.Regexp, s string) map[string]string {
+	matches := r.FindStringSubmatch(s)
+	capMap := make(map[string]string)
+	for i, name := range r.SubexpNames() {
+		if name != "" {
+			capMap[name] = matches[i]
+		}
+	}
+
+	return capMap
+}
+
+func mustFprintf(w io.Writer, format string, a ...interface{}) {
+	_, err := fmt.Fprintf(w, format, a...)
+	if err != nil {
+		panic("Assertion failed: mustFprintf")
+	}
+}
```

Comment:
Why did we have to inline all of this?  Would love to not have to do that in all repos.

Reply:
Because otherwise we would need a Go module for what is effectively a script which should be self contained in order to pull in pulumi/pulumi. We _could_ do this but given the last time it changed it seems to not be worthwhile vs inlining.

Reply:
I'd really love not to duplicate more boilerplate into these provider repos.  Can we go the other direction and just build this once out of a shared `scripts` project?  It does not appear to have any relationship to this repo.

Reply:
Yes, ideally this would live in scripts and be a binary dependency in the same way as `tf2pulumi` already is and `gomod-doccopy` was in the past. Right now I'm trying to make the minimal set of changes to get Go modules into the SDKs though, rather than wholesale refactoring of the build and versioning system!